src/dc_mimeparser.rs

```diff
@@ -422,22 +382,16 @@ unsafe fn dc_mimepart_new() -> dc_mimepart_t {
     }
 }
 
-pub unsafe fn dc_mimeparser_get_last_nonmeta(mimeparser: &dc_mimeparser_t) -> *mut dc_mimepart_t {
-    if !(*mimeparser).parts.is_null() {
-        let mut i: libc::c_int;
-        let icnt: libc::c_int = carray_count((*mimeparser).parts) as libc::c_int;
-        i = icnt - 1i32;
-        while i >= 0i32 {
-            let part: *mut dc_mimepart_t =
-                carray_get(mimeparser.parts, i as libc::c_uint) as *mut dc_mimepart_t;
-            if !part.is_null() && 0 == (*part).is_meta {
-                return part;
-            }
-            i -= 1
+pub unsafe fn dc_mimeparser_get_last_nonmeta<'a>(
+    mimeparser: &'a mut dc_mimeparser_t,
+) -> Option<&'a mut dc_mimepart_t> {
+    for part in mimeparser.parts.iter_mut().rev() {
+        if part.is_meta == 0 {
+            return Some(part);
         }
     }
 
-    0 as *mut dc_mimepart_t
+    None
```

Comment:
- I think this function is safe now
- it can be written as

```rust
pub unsafe fn dc_mimeparser_get_last_nonmeta<'a>(
    mimeparser: &'a mut dc_mimeparser_t,
) -> Option<&'a mut dc_mimepart_t> {
	mimeparser.parts.iter().rev().find(|part| part.is_meta == 0)
}
```

Reply:
Fixed